### ISSUE TITLE
Add tracing to functions with a context.Context

### ISSUE THREAD

**Issue by tbpg** [gomods/athens#324]
See #174, which threads contexts throughout the project.

Starter list (there are more outside of `storage`):
- [x] `storage.Exists`
- [x] `storage.Delete`
- [x] `storage.Get`
- [x] `storage.List`
- [x] `storage.Save`

From @arschles on #174:

> Here's an example on how to add a tracing span to a function:
> 
> ```go
> func foo(ctx buffalo.Context) {
>     sp, ctx := buffet.ChildSpan("storage.save", ctx)
>     defer sp.Close()
>     // do stuff with ctx
> }
> ```

----------------

I'll start with `List` and `Save`.

**Comment by manugupt1** [comment]
Hi @tbpg, I had like to take a look at this issue!

**Comment by tbpg** [comment]
`buffet` works with `buffalo.Context` for requests. So, we'll need to use the tracing library directly to add the spans. Should we use `opencensus` or `opentracing`? `buffet` seems to use `opentracing`, so should we continue with that?

Here's how to create an `opentracing` span:
```go
span, ctx := opentracing.StartSpanFromContext(ctx, "fs.List")
defer span.Finish()
```

Here's how to create an `opencensus` span:
```go
ctx, span := trace.StartSpan(ctx, "cache.Get")
defer span.End()
```

The APIs differ in how to create tracers, tags, & metrics, but I haven't used either extensively. If there hasn't been explicit decision, it might be good to split the tracing library question into a separate issue.

**Comment by arschles** [comment]
Please do!

**Comment by robjloranger** [comment]
:+1: for using opentracing until we fully convert to threading the buffalo context.

**Comment by manugupt1** [comment]
Got started on it here: https://github.com/gomods/athens/pull/326

The GET function touches a lot of other backends as well as the context was not defined earlier, so I am going to hold it off for a different commit.

**Comment by arschles** [comment]
@robjloranger is there an issue for fully convert to threading the buffalo context? If not, after #326 is merged, we should start talking about that in here

**Comment by robjloranger** [comment]
#174

**Comment by manugupt1** [comment]
@tbpg Just finished all of the items on the check list.
I have added a few more in this PR: https://github.com/gomods/athens/pull/332

Would you like me to just add for everything else in the storage package like GCP?

**Comment by manugupt1** [comment]
There is another one remaining which is the olympus backend?

**Comment by manugupt1** [comment]
Ref: https://github.com/gomods/athens/issues/392

**Comment by marpio** [comment]
@manugupt1 @tbpg Can we close this issue?

**Comment by manugupt1** [comment]
@marpio No :( There is a bug in the tracing thing. Right now, I am reading up on integrating about opencensus since we decided to use opencensus. 

I am on vacation this Friday - Monday, I am planning to write code after that. If anyone is willing to do it before that, it will be great too.

**Comment by manugupt1** [comment]
@arschles @marpio This issue was fixed with #604
